src/BaseMailer.php

```diff
@@ -0,0 +1,247 @@
+<?php
+namespace Yiisoft\Mailer;
+
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Psr\Log\LoggerInterface;
+use Yiisoft\Mailer\Event\{AfterSend, BeforeSend};
+
+/**
+ * BaseMailer serves as a base class that implements the basic functions required by [[MailerInterface]].
+ *
+ * Concrete child classes should may focus on implementing the [[sendMessage()]] method.
+ */
+abstract class BaseMailer implements MailerInterface
+{
+    /**
+     * @var bool whether to save email messages as files under [[fileTransportPath]] instead of sending them
+     * to the actual recipients. This is usually used during development for debugging purpose.
+     * @see fileTransportPath
+     */
+    public $useFileTransport = false;
```

Comment:
We need to get rid of public fields. All mandatory ones should become constructor arguments. All non-mandatory ones are to become setters.